bundle/src/main/java/com/adobe/acs/commons/ondeploy/impl/OnDeployExecutorImpl.java

```diff
@@ -0,0 +1,233 @@
+import org.apache.felix.scr.annotations.ConfigurationPolicy;
+import org.apache.felix.scr.annotations.Reference;
+import org.apache.felix.scr.annotations.ReferenceCardinality;
+import org.apache.felix.scr.annotations.ReferencePolicy;
+import org.apache.felix.scr.annotations.Service;
+import org.apache.sling.api.resource.LoginException;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.apache.sling.api.resource.ResourceResolverFactory;
+import org.apache.sling.commons.classloader.DynamicClassLoaderManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.jcr.Value;
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A service that triggers scripts on deployment to an AEM server.
+ * <p>
+ * This class manages scripts so that they only run once (unless the script
+ * fails).  Script execution statuses are stored in the JCR @
+ * /var/acs-commons/on-deploy-scripts-status.
+ * <p>
+ * Scripts are specified via OSGi config, are are run in the order specified.
+ * <p>
+ * NOTE: Since it's always a possibility that
+ * /var/acs-commons/on-deploy-scripts-status will be deleted in the JCR,
+ * scripts should be written defensively in case they are actually run more
+ * than once.  This also covers the scenario where a script is run a second
+ * time after failing the first time.
+ */
+@Component(
+        label = "ACS AEM Commons - On-Deploy Scripts Executor",
+        description = "Developer tool that triggers scripts (specified via an implementation of OnDeployScriptProvider) to execute on deployment.",
+        immediate = true,
+        metatype = true,
+        policy = ConfigurationPolicy.REQUIRE
+)
+@Service
+public class OnDeployExecutorImpl implements OnDeployExecutor {
+    private static final String SCRIPT_DATE_END = "endDate";
+    private static final String SCRIPT_DATE_START = "startDate";
+    private static final String SCRIPT_STATUS = "status";
+    private static final String SCRIPT_STATUS_JCR_FOLDER = "/var/acs-commons/on-deploy-scripts-status";
+    private static final String SCRIPT_STATUS_FAIL = "fail";
+    private static final String SCRIPT_STATUS_RUNNING = "running";
+    private static final String SCRIPT_STATUS_SUCCESS = "success";
+
+    private static final String SERVICE_NAME = "on-deploy-scripts";
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    @Reference
+    private ResourceResolverFactory resourceResolverFactory;
+    @Reference
+    private DynamicClassLoaderManager dynamicClassLoaderManager;
+    @Reference(name = "scriptProvider", referenceInterface = OnDeployScriptProvider.class, cardinality = ReferenceCardinality.MANDATORY_MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    private List<OnDeployScriptProvider> scriptProviders;
```

Comment:
I don't see an implementation of this interface in the PR?

Reply:
Correct.  This is a framework that allows other bundles to deploy scripts that will be run by this service.  ACS Commons would likely never include scripts itself.

Reply:
Ah, I thought the DeployScripts had to be implemented, it wasn't clear that also the provider had to be implemented :)

Reply:
Yes this allows different bundles to each provide scripts.  The reason the scripts are provided from coded OnDeployScriptProvider implementations rather than simply an OSGi config is to prevent a race condition with OSGi configs deploying separate from the script classes.

The provider class is pretty simple:

```
/**
 * Provider for the OnDeployExecutor service.
 *
 * Specifies a list of OnDeployScript instances to run.
 */
@Component(immediate = true)
@Service
@Properties({
        @Property(name = "service.description", value = "Developer service that identifies code scripts to run on deployment via the OnDeployExecutor service")
})
public class OnDeployScriptProviderImpl implements OnDeployScriptProvider {
    /**
     * Get the list of scripts to run.
     *
     * Ideally, all scripts will remain in the list indefinitely, so that a new AEM
     * server can run all scripts from the first to the last to be
     * completely up to date w/no manual intervention.  As a reminder, scripts will
     * run only once, so it is safe to preserve the entire list of scripts.
     *
     * @return List of OnDeployScript instances.
     */
    @Override
    public List<OnDeployScript> getScripts() {
        return Arrays.asList(
                new TestScript(),
                new TestScript2()
        );
    }
}
```

Reply:
I am getting unsatisfied reference error using it with 6.3 :+1: 
Unsatisfied
Service Name: com.adobe.acs.commons.ondeploy.OnDeployScriptProvider
Cardinality: 1..n
Policy: dynamic
Policy Option: reluctant
No Services bound

Reply:
@yuvraj1610 what does your implementation class for `OnDeployScriptProvider` look like?